tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/firebase/CassandraFirebaseSubscriptionRepository.scala

```diff
@@ -0,0 +1,84 @@
+package com.linagora.tmail.james.jmap.firebase
+
+import java.time.{Clock, ZonedDateTime}
+import java.util
+import java.util.Optional
+
+import com.linagora.tmail.james.jmap.firebase.FirebaseSubscriptionHelper.{evaluateExpiresTime, isInThePast, isNotOutdatedSubscription}
+import com.linagora.tmail.james.jmap.model.{DeviceClientIdInvalidException, DeviceTokenInvalidException, ExpireTimeInvalidException, FirebaseSubscription, FirebaseSubscriptionCreationRequest, FirebaseSubscriptionExpiredTime, FirebaseSubscriptionId, FirebaseSubscriptionNotFoundException}
+import javax.inject.Inject
+import org.apache.james.core.Username
+import org.apache.james.jmap.api.model.TypeName
+import org.reactivestreams.Publisher
+import reactor.core.publisher.SynchronousSink
+import reactor.core.scala.publisher.SMono
+
+import scala.jdk.javaapi.{CollectionConverters, OptionConverters}
+
+class CassandraFirebaseSubscriptionRepository @Inject()(dao: CassandraFirebaseSubscriptionDAO, clock: Clock) extends FirebaseSubscriptionRepository {
+
+  override def save(username: Username, request: FirebaseSubscriptionCreationRequest): Publisher[FirebaseSubscription] = {
+    val firebaseSubscription = FirebaseSubscription.from(request, evaluateExpiresTime(OptionConverters.toJava(request.expires.map(_.value)), clock))
+
+    isDuplicatedDeviceClientId(username, request.deviceClientId.value)
+      .handle((isDuplicatedDeviceClientId: Boolean, sink: SynchronousSink[AnyRef]) => {
+        if (isInThePast(request.expires, clock)) {
+          sink.error(ExpireTimeInvalidException(request.expires.get.value, "expires must be greater than now"))
+        }
+        if (isDuplicatedDeviceClientId) {
+          sink.error(DeviceClientIdInvalidException(request.deviceClientId, "deviceClientId must be unique"))
+        }})
+      .`then`(isDuplicatedDeviceToken(username, request.token.value))
+      .handle((isDuplicatedDeviceToken: Boolean, sink: SynchronousSink[AnyRef]) => {
+        if (isDuplicatedDeviceToken) {
+          sink.error(DeviceTokenInvalidException("deviceToken must be unique"))
+        }})
+      .`then`(dao.insert(username, firebaseSubscription))
+      .`then`(SMono.just(firebaseSubscription))
```

Comment:
thenReturn ?

Reply:
There is no thenReturn in scala reactor :p
![image](https://user-images.githubusercontent.com/55171818/199644307-60758df1-2127-4373-a2c2-3c3e65ee4c93.png)